Make date parsing more flexible for linedocsfile (europarl, enwiki)

dweiss opened the pull request:
Fixes #13073.

dweiss committed:
Make date parsing more flexible for linedocsfile (europarl, enwiki). #13073

dweiss commented:
This makes date parsing accept both europarl and enwiki. The tests still assume the docs come from europarl though and fail on the large enwiki.random.lines.txt:
```
gradlew -p lucene\backward-codecs test -Dtests.linedocsfile=/../enwiki.random.lines.txt
```

dweiss committed:
Remove the only assertion that holds for europarl only (?). There is another plain term query right below it (body:the) so I don't think it hurts much. #13073

uschindler reviewed (approved):
Looks fine. Why did you create a `Function<String,LocalDateTime>` for parsing instead of a simple static method? I think you did this to hide the formatter instances?

uschindler commented on lucene/backward-codecs/src/test/org/apache/lucene/backward_index/TestIndexSortBackwardsCompatibility.java:
```diff
@@ -147,6 +150,36 @@ public void testSortedIndex() throws Exception {
+        @Override
+        public LocalDateTime apply(String s) {
+          if (s.matches("^[0-9]{4}-[0-9]{2}-[0-9]{2}$")) {
+            return euroParl.parse(s, LocalDate::from).atTime(LocalTime.MIDNIGHT);
```
`.atStartOfDay()` would be shorter.

uschindler commented:
P.S.: Maybe add a quick test for the parsing logic to check that both formats are accepted.

dweiss commented:
> Looks fine. Why did you create a `Function<String,LocalDateTime>` for parsing instead of a simple static method? I think you did this to hide the formatter instances?

Yes, correct.

dweiss commented:
> P.S.: Maybe add a quick test for the parsing logic to check that both formats are accepted.

Maybe it'd be better to move this logic into LineFileDocs so that the date field's value is normalized somehow? Then we can add TestLineFileDocs and test it there. Adding a test method to TestIndexSortBackwardsCompatibility will multiply it over all the parameters, which doesn't make sense.

dweiss committed:
Use atStartOfDay().

dweiss committed:
Moved DATE_FIELD_VALUE_TO_LOCALDATETIME to LineFileDocs, added a test case to TestLineFileDocs

dweiss commented:
I've moved that utility function to LineFileDocs and added a basic test case to TestLineFileDocs (good idea). I feel tempted to normalize the date field's value in LineFileDocs but this may break other stuff so better to leave it for a follow-up issue.

dweiss commented:
I'll merge this in so that we can avoid jenkins failures. If there has to be a follow-up, I'll open another issue.

s1monw commented:
LGTM thanks
